File: libs/engine/src/java/platform.cc
#include "java/platform.h"

#include <stdexcept>

#include <hestia/engine/java.h>

namespace hestia::engine {
    namespace fs = std::filesystem;

    JavaTarget host_target() {
#if defined(_WIN32)
        const char *os = "windows";
#elif defined(__APPLE__)
        const char *os = "mac";
#else
        const char *os = "linux";
#endif
#if defined(__aarch64__) || defined(_M_ARM64)
        const char *arch = "aarch64";
#elif defined(__x86_64__) || defined(_M_X64)
        const char *arch = "x64";
#else
        const char *arch = nullptr;
#endif
        if (arch == nullptr) {
            throw std::runtime_error("no Java builds are published for this CPU architecture");
        }
        return JavaTarget{.os = os, .arch = arch};
    }

    namespace {
#if defined(_WIN32)
        constexpr const char *kJavaExe = "java.exe";
#else
        constexpr const char *kJavaExe = "java";
#endif

        std::optional<fs::path> java_under(const fs::path &home) {
            std::error_code ec;
            if (fs::path exe = home / "bin" / kJavaExe; fs::is_regular_file(exe, ec)) return exe;
            if (fs::path exe = home / "Contents" / "Home" / "bin" / kJavaExe; fs::is_regular_file(exe, ec)) {
                return exe;
            }
            return std::nullopt;
        }
    } // namespace

    std::optional<fs::path> find_java_executable(const fs::path &root) {
        if (auto exe = java_under(root)) return exe;
        std::error_code ec;
        for (const auto &entry: fs::directory_iterator(root, ec)) {
            if (!entry.is_directory(ec)) continue;
            if (auto exe = java_under(entry.path())) return exe;
        }
        return std::nullopt;
    }
} // namespace hestia::engine

File: crates/engine/src/java/platform.rs
//! Host platform mapping (Adoptium's vocabulary) and locating the java binary
//! inside an extracted archive.

use std::path::{Path, PathBuf};

use anyhow::{bail, Result};

/// Adoptium's platform vocabulary: os and architecture.
pub struct JavaTarget {
    pub os: String,
    pub arch: String,
}

pub fn host_target() -> Result<JavaTarget> {
    let os = if cfg!(target_os = "windows") {
        "windows"
    } else if cfg!(target_os = "macos") {
        "mac"
    } else {
        "linux"
    };
    let arch = if cfg!(target_arch = "aarch64") {
        "aarch64"
    } else if cfg!(target_arch = "x86_64") {
        "x64"
    } else {
        bail!("no Java builds are published for this CPU architecture");
    };
    Ok(JavaTarget {
        os: os.to_string(),
        arch: arch.to_string(),
    })
}

fn java_exe() -> &'static str {
    if cfg!(windows) {
        "java.exe"
    } else {
        "java"
    }
}

fn java_under(home: &Path) -> Option<PathBuf> {
    let direct = home.join("bin").join(java_exe());
    if direct.is_file() {
        return Some(direct);
    }
    let macos = home
        .join("Contents")
        .join("Home")
        .join("bin")
        .join(java_exe());
    if macos.is_file() {
        return Some(macos);
    }
    None
}

/// Find the java executable under an extracted archive: at the root, or one
/// directory level down (the usual `jdk-XX/` wrapper).
pub fn find_java_executable(root: &Path) -> Option<PathBuf> {
    if let Some(exe) = java_under(root) {
        return Some(exe);
    }
    let entries = std::fs::read_dir(root).ok()?;
    for entry in entries.flatten() {
        if entry.path().is_dir() {
            if let Some(exe) = java_under(&entry.path()) {
                return Some(exe);
            }
        }
    }
    None
}
